### PR TITLE
Remove assumption that infra replicas default to 2

### PR DESCRIPTION
**Pull request by jean-edouard**
**What this PR does / why we need it**:
When KUBEVIRT_INFRA_REPLICAS is not set, we still add `replicas: 2` in the CR, since it was the default until recently, and made the code simpler.
However, replicas are now dynamic, and we don't want to force a specific value when none is set.
This PR adds an `if` block to the CR template to address that.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Review by EdDev — COMMENTED** [review]
Can you please add the reasoning from the PR description to the commit message?

File: manifests/generated/kubevirt-cr.yaml.in [hunk 1]
```diff
@@ -14,6 +14,6 @@ spec:
       {{- end}}{{else}} []{{end}}
   customizeComponents: {}
   imagePullPolicy: {{.ImagePullPolicy}}
-  infra:
-    replicas: {{.InfraReplicas}}
+  infra:{{if .InfraReplicas}}
+    replicas: {{.InfraReplicas}}{{end}}
```
Comment (iholder-redhat) [inline manifests/generated/kubevirt-cr.yaml.in:18]: I admit to not really understand the changes here :)
Can you elaborate? What does the `if .InfraReplicas` refers to?

Thanks!

File: pkg/virt-operator/resource/generate/components/crds.go [hunk 2]
```diff
@@ -637,10 +635,8 @@ func NewKubeVirtCR(namespace string, pullPolicy corev1.PullPolicy, featureGates
 		}
 	}
 
-	if infraReplicas != util.DefaultInfraReplicas {
-		cr.Spec.Infra = &virtv1.ComponentConfig{
-			Replicas: &infraReplicas,
-		}
+	cr.Spec.Infra = &virtv1.ComponentConfig{
+		Replicas: &infraReplicas,
```
Comment (iholder-redhat) [inline pkg/virt-operator/resource/generate/components/crds.go:639]: I'm also confused by this.
[Here](https://github.com/kubevirt/kubevirt/blob/main/tools/resource-generator/resource-generator.go#L61), if `--infraReplicas` is not specified, we pass the `infraReplicas` as a placeholder (which equals `255`).

Why is this needed at all?
Why not do something like:
```go
if infraReplicas != 0 {
    cr.Spec.Infra = &virtv1.ComponentConfig{
        Replicas: &infraReplicas,
    }
}
```

Thanks

**Review by iholder-redhat — COMMENTED** [review]
Thanks a lot @jean-edouard!

Left a couple of questions :)

File: tools/resource-generator/resource-generator.go [hunk 3]
```diff
@@ -86,7 +86,11 @@ $1{{- end}}{{else}} []{{end}}`)
 	// However, when creating a template, we want its value to be something like "{{.InfraReplicas}}", which is not a uint8.
 	// Therefore, the value was substituted for a placeholder above (255). Replacing with the templated value now.
 	if strings.HasPrefix(*infraReplicasFlag, "{{") {
```
[the remark this answers is not in the document]
Comment (jean-edouard) [reply tools/resource-generator/resource-generator.go:88]: Hi, welcome to the rabbit hole, let me show you around! :)
For the CR, we only use `resource-generator` to generate a template (`kubevirt-cr.yaml.in`), which will then be translated to proper yaml by another tool. It's called from `hack/generate.sh` [1]
`resource-generator` could also be used to generate the final CR, and I suspect it is used that way by external projects wanting to integrate KubeVirt, hence the `if`.
The problem with using `resource-generator` to generate templates, is that the tool uses `yaml.Marshall()`, which doesn't really make sense since a template can't be expected to be proper yaml. However that wasn't a problem before we made feature gates (and then infra replicas) configurable, since all values were strings [2], and we were using simple `"{{.Bla}}"` templated values, which happen to be strings. Now that we have more complex (and potentially empty) values, we need to use placeholders first and mess with the text after the marshalling...

[1] https://github.com/kubevirt/kubevirt/blob/09445bf84b9213bb44c150b470ec9b88fbe788dd/hack/generate.sh#L138
[2] https://github.com/kubevirt/kubevirt/blob/09445bf84b9213bb44c150b470ec9b88fbe788dd/tools/manifest-templator/manifest-templator.go#L69

File: manifests/generated/kubevirt-cr.yaml.in (hunk 1 above)
Comment (jean-edouard) [reply manifests/generated/kubevirt-cr.yaml.in:18]: This is generated code from above (yes, generated template, meta!).
`if .InfraReplicas` will be true if this value is non-zero: https://github.com/kubevirt/kubevirt/blob/09445bf84b9213bb44c150b470ec9b88fbe788dd/tools/manifest-templator/manifest-templator.go#L70

File: pkg/virt-operator/resource/generate/components/crds.go (hunk 2 above)
Comment (jean-edouard) [reply pkg/virt-operator/resource/generate/components/crds.go:639]: In upstream, I believe `infraReplicas` will always be `255` here, since we also only use this for templates.
However, it could be used for other purposes. We could sanity check infraReplicas and panic if it's 0, but just adding that `if` might cause confusion...

File: tools/resource-generator/resource-generator.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (iholder-redhat) [reply tools/resource-generator/resource-generator.go:88]: Well, indeed a rabbit hole of insanity :D
Thanks for the explanation!

**Comment by iholder-redhat** [comment]
Thanks a lot @jean-edouard!!
/lgtm

File: pkg/virt-operator/resource/generate/components/crds.go [hunk 4]
```diff
@@ -637,10 +635,8 @@ func NewKubeVirtCR(namespace string, pullPolicy corev1.PullPolicy, featureGates
 		}
 	}
 
-	if infraReplicas != util.DefaultInfraReplicas {
-		cr.Spec.Infra = &virtv1.ComponentConfig{
-			Replicas: &infraReplicas,
-		}
+	cr.Spec.Infra = &virtv1.ComponentConfig{
```
Comment (xpivarc) [inline pkg/virt-operator/resource/generate/components/crds.go:638]: Maybe we should move this function closer to the user?

**Review by xpivarc — COMMENTED** [review]
/approve
/hold
@dhiller Can you have a look?

**Comment by dhiller** [comment]
@xpivarc what was the reason for the hold? Was it the stability of the lanes?
